### PR TITLE
add simple identity registry and registry-based authorizer

### LINKED ISSUE #1976: Add identity registry support for service accounts registered via `istioctl register`

### PR DESCRIPTION
**Pull request by crazytan**
**What this PR does / why we need it**:
This PR adds identity registry support for service accounts registered via `istioctl register` command. `istioctl register` has `-s` option to specify the service account to be linked to the service. By adding a service controller to listen to such changes (see https://github.com/istio/istio/pull/1951), the identity registry automatically adds new service accounts to the mapping (effectively a whitelist).

It also adds `registryAuthorizor` that consults identity registry for authorization decisions. When used by CA, the identity in the CSR sent by node agent needs to be in the registry to be signed by CA.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #1976

**Special notes for your reviewer**: There's no integration tests for this. Need to wait for https://github.com/istio/istio/issues/625. Also it's not enabled by default yet.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: security/pkg/registry/registry.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+// Registry is the standard interface for identity registry
+// implementation
+type Registry interface {
+	Check(string, string) bool
+	AddMapping(string, string)
+	DeleteMapping(string, string)
+}
+
+// IdentityRegistry is a naive registry that maintains a mapping between
+// identities (as strings)
```
Comment (wattli) [inline security/pkg/registry/registry.go]: Explains a little bit more about the mapping between identites?

File: security/pkg/registry/registry.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+func GetIdentityRegistry() Registry {
+	if reg == nil {
+		reg = &IdentityRegistry{
+			Map: make(map[string]string),
+		}
+	}
+	return reg
+}
+
+// K8SServiceAdded ...
```
Comment (wattli) [inline security/pkg/registry/registry.go]: Finish the comment? Same below

File: security/pkg/registry/registry.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	Map map[string]string
+}
+
+// Check checks whether id1 is mapped to id2
+func (reg *IdentityRegistry) Check(id1, id2 string) bool {
+	reg.RLock()
+	mapped, ok := reg.Map[id1]
+	reg.RUnlock()
+	if !ok {
+		return false
```
Comment (wattli) [inline security/pkg/registry/registry.go]: Can we log some info here for debugging purpose? And if possible, please also log the case that mapped existed, but not equal to id2.

File: security/pkg/server/grpc/authorizer.go [hunk 4]
```diff
@@ -47,3 +49,26 @@ func (authZ *sameIDAuthorizer) authorize(requester *caller, requestedIDs []strin
 
 	return nil
 }
+
+// registryAuthorizor uses an underlying identity registry to make authorization decisions
+type registryAuthorizor struct {
```
Comment (wattli) [inline security/pkg/server/grpc/authorizer.go]: So there will be another PR to hook up the Control Plane with this registryAuthorizor, right?

File: security/pkg/registry/registry.go (hunk 3 above)
Comment (crazytan) [reply security/pkg/registry/registry.go]: Will do.

File: security/pkg/registry/registry.go (hunk 2 above)
Comment (crazytan) [reply security/pkg/registry/registry.go]: sorry I forgot... Will add comments

File: security/pkg/server/grpc/authorizer.go (hunk 4 above)
Comment (crazytan) [reply security/pkg/server/grpc/authorizer.go]: I'm not sure whether to add a cmd option to let user select which authorizor to use, or just switch to `registryAuthorizor` by default (for this I need to ensure that existing mesh expansion script works fine).

File: security/pkg/registry/registry.go (hunk 1 above)
Comment (crazytan) [reply security/pkg/registry/registry.go]: Will do.

File: security/pkg/server/grpc/authorizer.go (hunk 4 above)
Comment (wattli) [reply security/pkg/server/grpc/authorizer.go]: The current one is pretty dummy and we should replace it. It would be great to provide an option for user to select once we have more complete authorizors. :)

Comment (wattli) [reply security/pkg/server/grpc/authorizer.go]: Can be done in a follow up PR :)
